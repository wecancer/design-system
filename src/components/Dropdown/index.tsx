import React, {useState, useRef} from 'react'
import styled, {css} from 'styled-components'

import OutsideEvent from '../OutsideEvent'

type AxisX = 'left' | 'right'

const Wrapper = styled(OutsideEvent)`
  position: relative;
  display: inline-block;
`

const TriggerContainer = styled.div``

const BoxContainer = styled.div<{triggerHeight: number; axisX: AxisX}>`
  ${({triggerHeight, axisX}) => css`
    position: absolute;
    top: ${triggerHeight + 10}px;
    ${axisX === 'left'
      ? css`
          left: 0;
        `
      : css`
          right: 0;
        `}
  `}
`

export type ActionParams = {
  isOpen: boolean
  handleClose(): void
  handleToggle(): void
}

export type Props = {
  axisX?: AxisX
  trigger(params: ActionParams): React.ReactNode
  children(params: ActionParams): React.ReactNode
  className?: string
}

const Dropdown = ({axisX = 'left', children, trigger, ...props}: Props) => {
  const refTrigger = useRef<HTMLDivElement>(null)
  const [isOpen, setOpen] = useState(false)
  const handleClose = () => setOpen(false)
  const handleToggle = () => setOpen(!isOpen)

  return (
    <Wrapper onClickOutside={handleClose} {...props}>
      <TriggerContainer ref={refTrigger}>
        {trigger({isOpen, handleClose, handleToggle})}
      </TriggerContainer>
      {isOpen && (
        <BoxContainer
          axisX={axisX}
          triggerHeight={refTrigger.current?.offsetHeight || 0}
        >
          {children({isOpen, handleClose, handleToggle})}
        </BoxContainer>
      )}
    </Wrapper>
  )
}

export default Dropdown
