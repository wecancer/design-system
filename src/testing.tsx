import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
