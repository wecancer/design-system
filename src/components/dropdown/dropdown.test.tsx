import Dropdown from '.'
import { render, fireEvent } from '../../testing'

describe('<Dropdown />', () => {
  it('should render snapshot correctly', () => {
    const { container } = render(
      <Dropdown
        className="custom-classname"
        trigger={() => <button type="button">Click me!</button>}
      >
        {() => <p>element hide</p>}
      </Dropdown>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should be show dropdown options when the trigger as clicked', () => {
    const { queryByTestId } = render(
      <Dropdown
        className="testing"
        trigger={({ handleToggle }) => (
          <button data-testid="trigger" type="button" onClick={handleToggle}>
            Click me!
          </button>
        )}
      >
        {({ handleClose }) => (
          <div data-testid="container">
            <button data-testid="op1" type="button">
              Option 1
            </button>
            <button data-testid="op2" type="button" onClick={handleClose}>
              Option 2
            </button>
            <button data-testid="op3" type="button">
              Option 3
            </button>
          </div>
        )}
      </Dropdown>,
    )
    const triggerButton = queryByTestId('trigger')
    fireEvent.click(triggerButton as HTMLElement)

    const option2 = queryByTestId('op2')

    expect(queryByTestId('container')).toBeDefined()
    fireEvent.click(option2 as HTMLElement)
    expect(queryByTestId('container')).toBeNull()
  })

  it('should be toggle the dropdown options when the trigger as clicked', () => {
    const { queryByTestId } = render(
      <Dropdown
        className="testing"
        trigger={({ handleToggle }) => (
          <button data-testid="trigger" type="button" onClick={handleToggle}>
            Click me!
          </button>
        )}
      >
        {({ handleClose }) => (
          <div data-testid="container">
            <button data-testid="op1" type="button">
              Option 1
            </button>
            <button data-testid="op2" type="button" onClick={handleClose}>
              Option 2
            </button>
            <button data-testid="op3" type="button">
              Option 3
            </button>
          </div>
        )}
      </Dropdown>,
    )

    const tiggerButton = queryByTestId('trigger') as HTMLElement

    fireEvent.click(tiggerButton)
    expect(queryByTestId('container')).toBeDefined()

    fireEvent.click(tiggerButton)
    expect(queryByTestId('container')).toBeNull()
  })
})
