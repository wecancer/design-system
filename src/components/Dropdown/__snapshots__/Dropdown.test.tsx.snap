// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render snapshot correctly 1`] = `
.c0 {
  position: relative;
  display: inline-block;
}

<div
  class="c0 custom-classname"
>
  <div
    class=""
  >
    <button
      type="button"
    >
      Click me!
    </button>
  </div>
</div>
`;
